### ISSUE TITLE
Rails master: after_commit(on: :update) is executed erroneously when a model is saved in an after_create callback

### ISSUE THREAD

**Issue by sidonath** [rails/rails#8937]
If a model is being saved in an `after_create` callback, then its internal `@_start_transaction_state[:new_record]` value will be set to `false`, indicating that the model was updated, thus hiding the fact that it was created as well.

This implies the `after_commit(on: :create)` will not be executed and `after_commit(on: :update)`—erroneously—will be.

```ruby
# user.rb
class User < ActiveRecord::Base
  after_create { self.save! }
  after_commit(on: :create) { puts "did run on create" }
  after_commit(on: :update) { puts "did run on update" }
end

# terminal
> User.create
   (0.1ms)  begin transaction
  SQL (4.9ms)  INSERT INTO "users" ("created_at", "updated_at") VALUES (?, ?)  [["created_at", Mon, 14 Jan 2013 17:46:24 UTC +00:00], ["updated_at", Mon, 14 Jan 2013 17:46:24 UTC +00:00]]
   (0.2ms)  UPDATE "users" SET "created_at" = '2013-01-14 17:46:24.710539', "updated_at" = '2013-01-14 17:46:24.710539' WHERE "users"."id" = 1
   (0.8ms)  commit transaction
did run on update
```

This change is breaking since 44d1804, which removed conditionals that prevented this kind of overlap.

The #8577 issue mentions another bug that is actually fixed with the commit 44d1804 (but can be fixed in a different way as well), so the fix is not to revert that commit, obviously.

/cc: @jjb

**Comment by jbedworth** [comment]
I'm seeing this exact behavior (calling self.save! in callback enters infinite loop) in rails 5 beta 3 in an after_create_commit callback.

**Comment by arthurnn** [comment]
I will take a look.

**Comment by jbedworth** [comment]
For what its worth, if I move my modifications to before_save, and don't call save! in after_create_commit, I can work around this issue.

**Comment by arthurnn** [comment]
I think this is strongly related to https://github.com/rails/rails/issues/14493, however the opposite.
If thats the case, I would say this is not a blocker for 5.0, as it is a known old miss behaviour.

I will remove the Milestone for now, and leave this open.
cc @rafaelfranca

**Comment by eugeneius** [comment]
The problem described in https://github.com/rails/rails/issues/8937#issuecomment-204641196 is actually https://github.com/rails/rails/issues/14493, and not this issue.

I think this can be closed, since it was fixed in v4.0.0 and hasn't regressed.
